cluster-operator/src/main/java/io/strimzi/operator/cluster/operator/resource/ZookeeperSetOperator.java

```diff
@@ -54,4 +102,225 @@ public static boolean needsRollingUpdate(StatefulSetDiff diff) {
         }
         return false;
     }
+
+    @Override
+    public Future<Void> maybeRollingUpdate(StatefulSet ss, Predicate<Pod> podRestart) {
+        String namespace = ss.getMetadata().getNamespace();
+        String name = ss.getMetadata().getName();
+        final int replicas = ss.getSpec().getReplicas();
+        log.debug("Considering rolling update of {}/{}", namespace, name);
+        Future<Void> f = Future.succeededFuture();
+
+
+        boolean zkRoll = false;
+        ArrayList<Pod> pods = new ArrayList<>();
+        String cluster = ss.getMetadata().getLabels().get(Labels.STRIMZI_CLUSTER_LABEL);
```

Comment:
I'm going to have a play at improving this algorithm. If I come up with something I think is better I'll open a PR against this branch and we can discuss it.

Reply:
Here's the algorithm I had in mind as pseudocode:

```
0. Create a list of `pods` initialised to `[0, replicas)`
1. While `size(pods) > 1`:
    1. Iterate over each of the numbers in `pods`:
        1. Fetch the pod for that pod number
        2. If the pod needs rolling:
            1. If the pod is the leader add it to the `leaders` list.
            2. If the pod is not the leader, then roll it and wait for readiness.
    // At this point we've rolled all the pods which needed rolling which were not the leader.
    // It is possible that a leader election happened while we did that though, 
    // so the `leaders` list could have 0 or more pods in it.
    2. `pods = leaders`
// At this point `pods` is either empty or has a single pod (the leader) in it.
2. If `pods` is non-empty then roll it and wait for readiness.
```

The reason I think this algo is better is that it copes with the possibility that a leadership election could happen spontaneously while we're rolling the non-leader pods.

I spent some time looking at this last night, but couldn't get something that was completely asynchronous (in the way I wanted). It's messy because of the while loop and (possibly) needing synchronization on the `pods` and `leaders`.

Reply:
> The reason I think this algo is better is that it copes with the possibility that a leadership election could happen spontaneously while we're rolling the non-leader pods.

Thinking about that some more I realised that there are two interesting cases, assuming that we're basically rolling in order of increasing pod number (except the leader):

1. The leader has a pod number > the pod we're currently rolling when a new leader is elected. If the new elected leader is a pod we've already rolled then the leader has rolled before the end and there's nothing we can do about that. Alternatively if the newly elected leader is a pod we'r yet to roll nothing has changed and we'll get to the new leader just as if it had always been leader.

2. If the leader has a pod number < the pod we're currently rolling when a new leader is elected (i.e. we've already not rolled what was the leader) then (assuming the leader was elected because it stopped responding) the SS controller should roll the old leader. Either the new leader is a pod we're yet to roll, in which case it doesn't make a difference so long as the algorithm resets what it thinks was the leader. Or the new leader is a pod we've already rolled, in which case it makes no different either.

So, long story short, I don't think we actually need the algorithm outlined above, some thing much more similar to what you already have will suffice.

What I would like to see though, is rather than iterating all the pods initially do decide whether a `zkRoll` is needed, if we just did the whole thing in one loop:

```
Future future = Future.succeddedFuture()
for podNum in [0, replicas):
    future.compose(ignored -> podOps.getAsync().compose(
        pod -> {
            if (podRestart.test(pod)) {
                if (isLeader(pod)) {
                    // record the leader
                } else {
                    // roll and wait for readiness
                }
            } else {
                return Future.succeededFuture();
            }
        );
}
future.compose(ignored -> {
    if (leader != null) { 
        //roll and wait on readiness
    }
});
```